dartagnan/src/main/java/com/dat3m/dartagnan/parsers/cat/VisitorCat.java

```diff
@@ -214,15 +206,30 @@ public Object visitExpr(ExprContext ctx) {
 
     @Override
     public Object visitExprNew(ExprNewContext ctx) {
-        return addDefinition(new Free(wmm.newRelation()));
+        final boolean unary = ctx.call.getText().equals("New");
+        return addDefinition(new Free(wmm.newRelation(unary)));
     }
 
     @Override
     public Object visitExprBasic(ExprBasicContext ctx) {
-        String name = ctx.n.getText();
-        Object predicate = namespace.computeIfAbsent(name,
-                k -> RelationNameRepository.contains(name) ? wmm.getOrCreatePredefinedRelation(k) : Filter.byTag(k));
-        return predicate;
+        final String name = ctx.n.getText();
+        final Object boundObject = namespace.get(name);
+        if (boundObject != null) {
+            return boundObject;
+        }
+        if (RelationNameRepository.contains(name)) {
+            return wmm.getOrCreatePredefinedRelation(name);
+        }
+        final Relation namedRelation = wmm.getRelation(name);
+        final Relation existingSet = namedRelation != null && namedRelation.getDefinition() instanceof TagSet s && s.getTag().equals(name) ?
+                namedRelation : wmm.getRelations().stream()
+                .filter(r -> r.getDefinition() instanceof TagSet s && s.getTag().equals(name))
+                .findAny().orElse(null);
+        if (existingSet != null) {
+            return existingSet;
+        }
```

Comment:
I feel all this is pointless. Just add the generated sets to the namespace (as the old code did) then this whole piece of code can get deleted. I guess this is also the reason why you needed to change "createUniqueName".